File: matlab_preprocessing/local_contrast_and_clahe.m
function [img_interp] = local_contrast_and_clahe(img,mask0)

ClipLimit = 0.005;
sigma = 80 / (55/25);
kernel = 250 / (55/25);



mask = imerode(mask0,strel('disk',1));

dt = bwdist(mask==1);


[sx,sy] = imgradientxy(dt,'sobel');

% magnitude = sqrt(sx.^2 + sy.^2);


[xx, yy] = meshgrid(1:size(dt,2), 1:size(dt,1));

xxx = xx - sx/4 .* dt;
yyy = yy - sy/4 .* dt;


img_interp(:,:,1) = interp2(img(:,:,1),xxx,yyy,'linear',0);
img_interp(:,:,2) = interp2(img(:,:,2),xxx,yyy,'linear',0);
img_interp(:,:,3) = interp2(img(:,:,3),xxx,yyy,'linear',0);

G = imgaussfilt(img_interp, sigma, 'Padding', 'replicate');
img_interp = (img_interp - G) ./ G + 0.5;
img_interp(img_interp < 0) = 0; 
img_interp(img_interp > 1) = 1;

img_interp = rgb2hsv(img_interp);

img_interp(:,:,3) = adapthisteq(img_interp(:,:,3), 'NumTiles',[round(size(img_interp,1) / kernel) round(size(img_interp,2) / kernel)],'ClipLimit',ClipLimit);

img_interp = hsv2rgb(img_interp);

img_interp(repmat(mask0 == 0,[1,1,3])) = 0;

File: utils/local_contrast_and_clahe.py
import numpy as np
import matplotlib.pyplot as plt
from skimage.io import imread
from scipy.ndimage import distance_transform_edt
from scipy.ndimage import sobel
import cv2
from scipy.ndimage import binary_erosion
from skimage.morphology import disk
from skimage.exposure import equalize_adapthist
from skimage.filters import gaussian
from skimage.color import rgb2hsv, hsv2rgb


def local_contrast_and_clahe(img, fov, pix_per_deg):

    ClipLimit = 0.005
    # original value chosen for 55 pix_per_deg
    sigma = 80 / (55 / pix_per_deg)
    kernel = 250 / (55 / pix_per_deg)
    
    img = img.astype(np.float32) / 255 
    

    mask = binary_erosion(fov, disk(1))

    dt = distance_transform_edt(mask == 0)

    sx = sobel(dt, axis=0) / 4
    sy = sobel(dt, axis=1) / 4

    xx, yy = np.meshgrid(np.arange(dt.shape[1]), np.arange(dt.shape[0]))

    xxx = xx - sy * dt
    yyy = yy - sx * dt

    img_interp = cv2.remap(img, xxx.astype(np.float32), yyy.astype(np.float32), cv2.INTER_LINEAR)

    G = gaussian(img_interp, sigma, channel_axis = 2)
    img_interp = (img_interp - G) / G + 0.5
    img_interp[img_interp < 0] = 0
    img_interp[img_interp > 1] = 1

    img_interp = rgb2hsv(img_interp)
    img_interp[:,:,2] = equalize_adapthist(img_interp[:,:,2], kernel_size=kernel, clip_limit=ClipLimit)
    img_interp = hsv2rgb(img_interp)

    img_interp[fov == 0] = 0
    
    img_interp = img_interp * 255 
    img_interp[img_interp < 0] = 0
    img_interp[img_interp > 255] = 255
    img_interp = img_interp.astype(np.uint8) 

    return img_interp



if __name__ == "__main__":
    
    import sys
    sys.path.append("..")
    from datasets_loaders.LoaderGeneric import LoaderGeneric
    
    img = imread(r"..\..\databases\ARIA\aria_a_markups\aria_a_markups\aria_a_5_9.tif")
    img_norm = local_contrast_and_clahe(img, LoaderGeneric.get_fov_auto(img), 50)
    
    plt.imshow(img)
    plt.show()
    
    plt.imshow(img_norm)
    plt.show()